data_management/rocrate.py

```diff
@@ -0,0 +1,661 @@
+    for component in object_components:
+        obj = component.object
+        data_products = obj.data_products.all()
+
+        for data_product in data_products:
+            crate_data_product = _get_data_product(
+                crate, data_product, registry_url, output
+            )
+            all_data_products.append(crate_data_product)
+
+    if output:
+        crate_code_run["result"] = all_data_products
+    else:
+        crate_code_run["object"] = all_data_products
+
+
+def _add_licenses(crate, crate_entity, file_object, registry_url):
+    """
+    Add licenses from the file_object to the crate_entity.
+
+    @param crate: the RO Crate object
+    @param crate_entity: an entity to add the license to
+    @param file_object: an "object" from the database representing a file
+    @param registry_url: a str containing the registry URL
+
+    """
+    license_entities = []
+    try:
+        licenses = file_object.licences.all()
+    except AttributeError:
+        licenses = []
+
+    for license_ in licenses:
+        if license_.identifier is not None:
+            license_id = license_.identifier
+        else:
+            license_id = f"{registry_url}api/license/{license_.id}"
+
+        license_entity = ContextEntity(
+            crate,
+            license_id,
+            properties={
+                RO_TYPE: "CreativeWork",
+                "description": license_.licence_info,
+                "identifier": license_id,
+                "name": f"license {license_.id}",
+            },
+        )
+
+        crate.add(license_entity)
+        license_entities.append(license_entity)
+
+    # where the crate_entity is a crate we are adding the licenses from all the
+    # data products in turn, so there may already be some there
+    if isinstance(crate_entity, ROCrate) and crate_entity.license is not None:
+        license_entities.extend(crate_entity.license)
+
+    if len(license_entities) == 1:
+        if isinstance(crate_entity, ROCrate):
+            crate_entity.license = license_entities[0]
+        else:
+            crate_entity["license"] = license_entities[0]
+
+    elif len(license_entities) > 1:
+        if isinstance(crate_entity, ROCrate):
+            crate_entity.license = license_entities
+        else:
+            crate_entity["license"] = license_entities
+
+
+def _add_metadata_license(crate):
+    """
+    Add a licenses to the ro-crate-metadata.json file.
+
+    @param crate: the RO Crate object
+
+    """
+    url = "https://creativecommons.org/publicdomain/zero/1.0/"
+    metadata_license = ContextEntity(
+        crate,
+        url,
+        properties={
+            RO_TYPE: "CreativeWork",
+            "description": "CC0 1.0 Universal (CC0 1.0) Public Domain Dedication",
+            "identifier": url,
+            "name": "CC0 Public Domain Dedication",
+        },
+    )
+
+    for entity in crate.default_entities:
+        if entity.id == "ro-crate-metadata.json":
```

Comment:
No need to loop. You can use `get` (or its alias, `dereference`) to get any entity by its id:

```python
crate.get("ro-crate-metadata.json")
```

Or, in this case:

```python
crate.metadata
```

So the loop can be replaced by:

```python
crate.add(metadata_license)
crate.metadata["license"] = metadata_license
```

Reply:
Updated in https://github.com/FAIRDataPipeline/data-registry/pull/189/commits/a3ecb9741f8e9649a083318f4dcc30e36615c1f6